_posts/blog/2016-10-08-vector_drawables_optimisation.md

```diff
@@ -0,0 +1,188 @@
+Vector drawables allow the representation of images (e.g. icons, UI elements) based on XML vector graphics.
+
+
+<center>
+<picture>
+	<a href="/images/blog/vector_drawables/vector_drawables_vs_png.png"><img src="/images/blog/vector_drawables/vector_drawables_vs_png.png" alt="VectorDrawable vs PNG" width="350"></a>
+	<figcaption>Using a VectorDrawable vs using a small PNG image</figcaption>
+</picture>
+</center>
+
+
+Using vector data instead of raster image data, the number and the size of resources added to the project decreases, since now only one resource per resolution is needed, and therefore also the APK size.
+
+
+Although the XML file containing the vector drawable is usually smaller than the PNG version, the vector drawables come with a computational overhead at runtime, which may be an issue for more complex graphical elements. When vector drawables are drawn for the first time, a cached bitmap is created in order to optimize the re-drawing performance. This cache is re-used as long as the width and the height of the image that needs to be drawn is the same. If a VectorDrawable is used for multiple sizes, a new Bitmap will be created every time.
+
+
+Compared to raster images, drawing VectorDrawables will take more time for the first rendering. But then, as long as the size of the image doesn't change, the next drawings of VectorDrawables will take a similar amount of time,  just like drawing raster images.
+
+
+
+
+## Multiple Sizes, Multiple Renderings
+
+
+Let's say that you need to display an image that needs to fill the height of the screen but keep the aspect ration when rotating the device from portrait to landscape.
+
+
+<center>
+<picture>
+	<a href="/images/blog/vector_drawables/phone_portrait.png"><img src="/images/blog/vector_drawables/phone_portrait.png" alt="VectorDrawable vs PNG" height="350"></a>
+	<a href="/images/blog/vector_drawables/phone_landscape.png"><img src="/images/blog/vector_drawables/phone_landscape.png" alt="VectorDrawable vs PNG" width="350"></a>
+	<figcaption>Changing the size of the image on orientation</figcaption>
+</picture>
+</center>
+
+
+**VectorDrawables use a Bitmap cache that gets recreated when the size changes**, so for example when you rotate your device from portrait to landscape. After the first rendering, the cached bitmap will be used. This means that you end up spending a lot of time on the first rendering of the VectorDrawable, at every rotation of the screen.
+The best approach, in this case,is to **create one VectorDrawable for portrait and one for landscape**. Like this, time will be spent only for the first rendering of the two images and the corresponding bitmap caches will be used from there on for every rotation.
+
+
+Let's test this! The size of the view changes when changing the orientation. When rotating the device, drawing the vector drawable takes, in average, 15.50ms in portrait and 7.80ms in landscape, where the view is smaller.
```

Comment:
I wouldn't use absolute numbers - isn't that dependent on the svg and the performance of the phone and so on?